### ISSUE TITLE
Add individual exit codes for different errors

### ISSUE THREAD

**Issue by Prinzhorn** [mitmproxy/mitmproxy#5696]
#### Problem Description

This came from https://github.com/mitmproxy/mitmproxy/issues/4669 and was attempted in https://github.com/mitmproxy/mitmproxy/pull/5133

Currently every error that causes mitmproxy to go down has an exit code of `1`. We should add more specific error codes.

#### Proposal

Replace all `sys.exit(1)` with more specific exit codes `>=1` so that a parent process gets more information why mitmproxy (likely `mitmdump`) went down.

Example codes I came up with in the failed PR attempt:

https://github.com/mitmproxy/mitmproxy/pull/5133/files#diff-eb2a1155058d5000102ff2c85ec2acb691ca8277539e2d1c9702dbbab583dfa9R1-R12

**Comment by redterrum** [comment]
@Prinzhorn I'm new to open-source and would love to work on this issue

**Comment by Prinzhorn** [comment]
Welcome! Sure, go ahead. Feel free to post any questions here. You can take a look at the changes made in https://github.com/mitmproxy/mitmproxy/pull/5133/files (which was never merged)

1. Define exit codes in a central place (see exit_codes.py)
2. Maybe you can come up with a better or different numbering scheme. I tried to group them somewhat logically and not just number them 1,2,3... because if later a new exit codes is added it can be added to the appropriate category instead of either at the end or having to move all others
3. Replace `sys.exit(1)` calls with the matching error codes
4. Add tests for each exit code to a `test_exit_codes.py` file (that's mostly why the other PR was never finished, I'm not good with Python :sweat_smile:  )
5. And then maybe find a place in the docs for it

**Comment by Prinzhorn** [comment]
Also https://github.com/mitmproxy/mitmproxy/blob/main/CONTRIBUTING.md contains all instructions needed to setup the local Python environment

**Comment by redterrum** [comment]
Thanks @Prinzhorn ! I'll get started right away

**Comment by redterrum** [comment]
Hi @Prinzhorn , I went through the code and while I'm done with the implementation. It's the testing code that is problematic for me. I believe this is beyond my current knowledge, and it would be better if someone more skilled took this up. Thanks!

**Comment by Prinzhorn** [comment]
If you want to you can still create a pull request and maybe we can figure everything out. In my original PR I had one test https://github.com/Prinzhorn/mitmproxy/blob/6d5db0dc6a97f0667b5a163b0f8e9ac5a3806b4b/test/mitmproxy/utils/test_exit_codes.py
It's mocking `print` to trigger the exit via a raise. Some of the `sys.exit(1)` code paths might already have tests and we just need to add a line that checks the proper code. But no worries if you can't finish it.

**Comment by kiklidash** [comment]
Hi @Prinzhorn , can I take a shot at this?

**Comment by Prinzhorn** [comment]
@kiklidash sure! I think everything that needs to be done is outlined above.

**Comment by kiklidash** [comment]
Hi @Prinzhorn , I'm done replacing the exit codes in the code base and with a couple of tests too. I just need help with testing for INVALID_ARGS and INVALID_OPTIONS in mitmproxy/tools/main.py and the exit codes in mitmproxy/tools/console/master.py
I took a look into the test scripts for these files, but it wasn't very helpful. Can you help me out with this? Thanks!

**Comment by Prinzhorn** [comment]
> Can you help me out with this?

I failed writing the tests as well :sweat_smile: . You can open a pull request with what you have and we can go from there. That makes it easier to give specific advice.

**Comment by kiklidash** [comment]
Hi, I've opened a [PR](https://github.com/mitmproxy/mitmproxy/pull/5882). You can go through my changes and we can take this forward from here on. Thanks!

**Comment by cynhash** [comment]
Is this still open?

**Comment by Prinzhorn** [comment]
Yes. The last two people attempting it deleted their account. All required information is there.

**Comment by cynhash** [comment]
I'll take a look at the PRs attached and see where I can start

**Comment by alexgershberg** [comment]
> Yes. The last two people attempting it deleted their account. All required information is there.

That's rather grim

**Comment by Prinzhorn** [comment]
@smenon02 everything outlined in https://github.com/mitmproxy/mitmproxy/issues/5696#issuecomment-1304439116

**Comment by pratik-ja** [comment]
@Prinzhorn 
I see that the issue is still open and everything that needs to be done is outlined. Will start working on this asap.
